vsr: quicker request protocol

matklad opened the pull request:
Old protocol:

- client round-robins the target replica during retries
- replica forwards a request to the primary _if_ the message is from the
  earlier view
- if the request is from the same view, it is dropped

This really doesn't work at all --- if a single message is dropped or
delayed, the client has to go through the entire ring unitl it re-sends
back to the primary.

New protocol:

- client always sends a message to a replica it thinks to be the primary.
- on retry, client sends a round-robin ping message to learn if a view
  change happened.
- backups never forward requests

Things intentionally omitted:

- When a client receives a pong and advances its view, it _does not_
  reset the on_request timeout. This is to spread out the thundering
  heard after a view change, when the cluster might have a backlog of
  prepares.
- Replicas never forward a request, as that is prone to amplify
  bandwidth when the cluster isn't fully healthy.
- Replicas _do not_ respond with a pong to a request with a wrong view.
  Because the client always sends request to the primary, and the
  primary might be down, it _has_ to send a ping to someone else, and
  it's that replica's job to inform the client about the view.
- The original `.register` request is not broadcast, and follows the same logic.
  This is a bit unintelligent, as client's initial view might be wrong,
  and broadcasting feels better. However, I am not sure our on_request
  timeout is set correctly: in local VOPR testing with very fast
  storage, somewhat faulty networks, and multiple clients, the clients
  often do useless retries, which lead to "request already
  queued/preparing" on the primary. I will look into the timeout
  separately, and, for now, use the same logic for .register as for all
  other messages, as that's just simpler.

Results for `./zig/zig build vopr -- --performance 92` with one client and 10% packet loss:

```
Before:

Messages:
ping                     10650   5.20MiB
pong                     9574    2.34MiB
commit                   3615    903.75KiB
prepare                  1563    578.34KiB
prepare_ok               1428    357.00KiB
request                  561     202.00KiB
reply                    280     106.84KiB
request_prepare          228     57.00KiB
request_headers          157     39.25KiB
headers                  139     501.00KiB
ping_client              66      16.50KiB
pong_client              59      14.75KiB
request_start_view       51      12.75KiB
start_view               51      138.50KiB
do_view_change           25      12.50KiB
total                    28447   10.41MiB

          PASSED (35075 ticks)
```

```
After:

Messages:
ping                     3210    1.57MiB
pong                     2890    722.50KiB
prepare                  1682    727.20KiB
prepare_ok               1518    379.50KiB
commit                   1140    285.00KiB
request                  348     153.91KiB
reply                    302     134.94KiB
request_prepare          273     68.25KiB
request_headers          164     41.00KiB
headers                  151     552.75KiB
ping_client              109     27.25KiB
pong_client              99      24.75KiB
start_view               38      101.50KiB
request_start_view       36      9.00KiB
do_view_change           25      12.50KiB
total                    11985   4.73MiB

          PASSED (10296 ticks)
```

Note how both the number of ticks, and the number of requests goes down,
as expected.

Six clients and no packet loss:

```
Before:

Messages:
prepare                  1390    617.47KiB
prepare_ok               1299    324.75KiB
ping                     630     315.00KiB
pong                     630     157.50KiB
request                  276     123.63KiB
commit                   275     68.75KiB
reply                    262     110.00KiB
request_headers          62      15.50KiB
headers                  60      190.25KiB
request_prepare          32      8.00KiB
do_view_change           30      15.00KiB
start_view               5       7.50KiB
total                    4951    1.91MiB

          PASSED (1652 ticks)
```

```
After:

Messages:
prepare                  1425    634.95KiB
prepare_ok               1420    355.00KiB
ping                     660     330.00KiB
pong                     660     165.00KiB
request                  309     141.88KiB
commit                   290     72.50KiB
reply                    277     112.72KiB
request_headers          71      17.75KiB
headers                  71      242.75KiB
ping_client              47      11.75KiB
pong_client              46      11.50KiB
do_view_change           30      15.00KiB
request_prepare          26      6.50KiB
start_view               5       7.50KiB
total                    5337    2.07MiB

          PASSED (1783 ticks)
```

I don't understand why we get more requests in this scenario, but my gut
feeling is that something's off with our on_request timeout, as we do get some
retries even in the perfect network with fast disk!

```
λ rg 'on_request: ignoring' log.vopr
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already queued)
[debug] (replica): 1: on_request: ignoring (pipeline full)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (pipeline full)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already queued)
[debug] (replica): 1: on_request: ignoring (pipeline full)
[debug] (replica): 1: on_request: ignoring (already queued)
[debug] (replica): 1: on_request: ignoring (pipeline full)
[debug] (replica): 1: on_request: ignoring (already queued)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (pipeline full)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already preparing)
```

I think this is what happening --- before, we finished the test without sending a single ping_client. After, we do send a ping_client on retry, and that lowers the rtt in the on_request timeout, and we start retrying more.

matklad committed:
vsr: new request retry protocol

Old protocol:

- client round-robins the target replica during retries
- replica forwards a request to the primary _if_ the message is from the
  earlier view
- if the request is from the same view, it is dropped

This really doesn't work at all --- if a single message is dropped or
delayed, the client has to go through the entire ring unitl it re-sends
back to the primary.

New protocol:

- client always sends a message to a replica it thinks to be the primary.
- on retry, client sends a round-robin ping message to learn if a view
  change happened.
- backups never forward requests

Thins intentionally omitted:

- When a backup receives a pong and advances its view, it _does not_
  reset the on_request timeout. This is to spread out the thundering
  heard after a view change, when the cluster might have a backlog of
  prepares.
- Replicas never forward a request, as that is prone to amplify
  bandwidth when the cluster isn't fully healthy.
- Replicas _do not_ respond with a pong to a request with a wrong view.
  Because the client always sends request to the primary, and the
  primary might be down, it _has_ to send a ping to someone else, and
  it's that replica's job to inform the client about the view.
- The original request is not broadcast, and follows the same logic.
  This is a bit unintelligent, as client's initial view might be wrong,
  and broadcasting feels better. However, I am not sure our on_request
  timeout is set correctly: in local VOPR testing with very fast
  storage, somewhat faulty networks, and multiple clients, the clients
  often do useless retries, which lead to "request already
  queued/preparing" on the primary. I will look into the timeout
  separately, and, for now, use the same logic for .register as for all
  other messages, as that's just simpler.

Results with one client and 10% packet loss:

```
Before:

Messages:
ping                     10650   5.20MiB
pong                     9574    2.34MiB
commit                   3615    903.75KiB
prepare                  1563    578.34KiB
prepare_ok               1428    357.00KiB
request                  561     202.00KiB
reply                    280     106.84KiB
request_prepare          228     57.00KiB
request_headers          157     39.25KiB
headers                  139     501.00KiB
ping_client              66      16.50KiB
pong_client              59      14.75KiB
request_start_view       51      12.75KiB
start_view               51      138.50KiB
do_view_change           25      12.50KiB
total                    28447   10.41MiB

          PASSED (35075 ticks)
```

```
After:

Messages:
ping                     3210    1.57MiB
pong                     2890    722.50KiB
prepare                  1682    727.20KiB
prepare_ok               1518    379.50KiB
commit                   1140    285.00KiB
request                  348     153.91KiB
reply                    302     134.94KiB
request_prepare          273     68.25KiB
request_headers          164     41.00KiB
headers                  151     552.75KiB
ping_client              109     27.25KiB
pong_client              99      24.75KiB
start_view               38      101.50KiB
request_start_view       36      9.00KiB
do_view_change           25      12.50KiB
total                    11985   4.73MiB

          PASSED (10296 ticks)
```

Note how both the number of ticks, and the number of requests goes down,
as expected.

Six clients and no packet loss:

```
Before:

Messages:
prepare                  1390    617.47KiB
prepare_ok               1299    324.75KiB
ping                     630     315.00KiB
pong                     630     157.50KiB
request                  276     123.63KiB
commit                   275     68.75KiB
reply                    262     110.00KiB
request_headers          62      15.50KiB
headers                  60      190.25KiB
request_prepare          32      8.00KiB
do_view_change           30      15.00KiB
start_view               5       7.50KiB
total                    4951    1.91MiB

          PASSED (1652 ticks)
```

```
After:

Messages:
prepare                  1425    634.95KiB
prepare_ok               1420    355.00KiB
ping                     660     330.00KiB
pong                     660     165.00KiB
request                  309     141.88KiB
commit                   290     72.50KiB
reply                    277     112.72KiB
request_headers          71      17.75KiB
headers                  71      242.75KiB
ping_client              47      11.75KiB
pong_client              46      11.50KiB
do_view_change           30      15.00KiB
request_prepare          26      6.50KiB
start_view               5       7.50KiB
total                    5337    2.07MiB

          PASSED (1783 ticks)
```

I don't understand why we get more requests in this scenario, but my gut
feeling is that something's off with our timeout, as we do get some
retries even in the perfect network with fast disk!

```
λ rg 'on_request: ignoring' log.vopr
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 0: on_request: ignoring (backup view=1 header.view=0)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already queued)
[debug] (replica): 1: on_request: ignoring (pipeline full)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (pipeline full)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already queued)
[debug] (replica): 1: on_request: ignoring (pipeline full)
[debug] (replica): 1: on_request: ignoring (already queued)
[debug] (replica): 1: on_request: ignoring (pipeline full)
[debug] (replica): 1: on_request: ignoring (already queued)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (pipeline full)
[debug] (replica): 1: on_request: ignoring (already preparing)
[debug] (replica): 1: on_request: ignoring (already preparing)
```

matklad committed:
vopr: warm up the cluster in performance mode

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -734,9 +764,7 @@ pub fn ClientType(
-            // If our view number is out of date, then the old primary will forward our request.
-            // If the primary is offline, then our request timeout will fire and we will
-            // round-robin.
+            // Otherwise we know the view and send directly to the primary.
```
This "Otherwise" feels like it is missing some context.

matklad replied:
Ah, that's the remnant of "broadcast .register" code which I decided to take out for now!

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -5523,7 +5523,17 @@ pub fn ReplicaType(
+            if (self.backup()) {
+                // Clients only send requests to the primary. Either the client's view is outdated,
+                // or a message was misdirected. Intentinonally don't try to forward the message to
+                // the primary, to avoid amplyfying the network load.
```
```suggestion
                // the primary, to avoid amplifying the network load.
```

sentientwaffle commented on src/vsr/message_header.zig:
```diff
@@ -201,7 +201,8 @@ pub const Header = extern struct {
         switch (self.into_any()) {
             .reserved => unreachable,
-            // These messages cannot always identify the peer as they may be forwarded:
+            // TODO: replicas used to forward requests. They no longer do, and can always return
+            // request.client starting with the next release.
```
Nice catch; I had forgotten about this!

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -451,6 +452,10 @@ pub fn ClientType(
                 self.view = pong.header.view;
+                // Even if there is a requst in flight, don't try to retransmit it immediately after
+                // a view change. Instead, ride the on_request_timeout normally  to reduce the size
+                // of thundering heard.
```
```suggestion
                // Even if there is a request in flight, don't try to retransmit it immediately after
                // a view change. Instead, ride the on_request_timeout normally to reduce the size
                // of thundering herd.
```

matklad replied:
Ahh, heard isn't a typo, I genuinely thought that's how you spell it, thanks!

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -5523,7 +5523,17 @@ pub fn ReplicaType(
             // This check must precede any send_eviction_message_to_client(), since only the primary
             // should send evictions.
-            if (self.ignore_request_message_backup(message)) return true;
+            if (self.backup()) {
```
Thinking out loud:

When a cluster running this code is combined with a client running the old code, the old client still works since even though its `register` won't be forwarded, it is round-robin-ing its requests, so it will stumble on the primary eventually. And also it has a 30s ping broadcast upper bound.

matklad replied:
>even though its register won't be forwarded

I think we never forwarded register in the first place? We did forward non-register requests from _older_ views, so what regresses is that, after a view-change, the client needs either to round-robin to the primary, or wait 30 seconds for a ping, whichever is faster.

sentientwaffle replied:
Ah right, I got it backwards which one we forwarded!

sentientwaffle reviewed (commented):
A couple of nits, but otherwise this looks good to me.

matklad committed:
vsr: code review

matklad committed:
vsr: remove incorrect timeout assert

timeout intentionally wraps around on u8, so it _can_ be zero in long
vopr runs.
